operator/internal/controller/console/controller.go

```diff
@@ -0,0 +1,331 @@
+// Copyright 2025 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+package console
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"reflect"
+	"time"
+
+	"github.com/cockroachdb/errors"
+	"github.com/imdario/mergo"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/utils/ptr"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/redpanda-data/redpanda-operator/charts/console/v3"
+	redpandav1alpha2 "github.com/redpanda-data/redpanda-operator/operator/api/redpanda/v1alpha2"
+	"github.com/redpanda-data/redpanda-operator/operator/api/redpanda/v1alpha2/conversion"
+	"github.com/redpanda-data/redpanda-operator/pkg/kube"
+)
+
+const (
+	managedByService      = "redpanda-operator"
+	clusterReferenceIndex = "__console_referencing_cluster"
+)
+
+// console resources
+// +kubebuilder:rbac:groups=cluster.redpanda.com,resources=consoles,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=cluster.redpanda.com,resources=consoles/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=autoscaling,resources=horizontalpodautoscalers,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=networking.k8s.io,resources=ingresses,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=core,resources=configmaps;secrets;services;serviceaccounts,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=apps,resources=deployments,verbs=get;list;watch;create;update;patch;delete
+
+type Controller struct {
+	Ctl *kube.Ctl
+
+	// rng is used to generate Console's JWT Signing keys, if they're not
+	// explicitly specified. If nil, SetupWithManager will set it with a seeded
+	// value.
+	rng *rand.Rand
+}
+
+func (c *Controller) SetupWithManager(ctx context.Context, mgr ctrl.Manager) error {
+	// If rng is not set for testing, create and seed a new one.
+	if c.rng == nil {
+		// TODO: Weak RNG is probably acceptable here but best to doublecheck
+		c.rng = rand.New(rand.NewSource(time.Now().UnixMicro())) //nolint:gosec
+	}
+
+	// Setup a field index for clusterRef so we can trigger updates to Consoles when Redpandas are changed.
+	if err := mgr.GetFieldIndexer().IndexField(ctx, &redpandav1alpha2.Console{}, clusterReferenceIndex, func(o client.Object) []string {
+		console := o.(*redpandav1alpha2.Console)
+		if console.Spec.ClusterSource == nil || console.Spec.ClusterSource.ClusterRef == nil {
```

Comment:
Technically without doing a `Kind`/`Group` check this is now wrong, see my other comment about leaning on `registerClusterSourceIndex` which would do those checks internally for you.

Reply:
Done.